### PR TITLE
Fix RefreshIndicator performance issue

### LINKED ISSUE #47664: RefreshIndicator is expensive
The `RefreshIndicator` is often used as a parent of a ListView. If the ListView is more complex the RefreshIndicator widget is very expensive performance-wise because the RefreshIndicator moves the child with a global key on a stack if the refresh indicator is to be shown. This causes a relayout and repaint of the whole subtree.

### PR DESCRIPTION
**Pull request by christian-muertz**
## Description

Fixes the performance issue described by https://github.com/flutter/flutter/issues/47664.

## Related Issues

-Fixes https://github.com/flutter/flutter/issues/47664

## Tests

I added the following tests: /

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

### CODE REVIEW

**Comment by christian-muertz** [comment]
@googlebot I fixed it.

File: packages/flutter/lib/src/material/refresh_indicator.dart [hunk 1]
```diff
@@ -421,18 +421,18 @@ class RefreshIndicatorState extends State<RefreshIndicator> with TickerProviderS
     if (_mode == null) {
       assert(_dragOffset == null);
       assert(_isIndicatorAtTop == null);
-      return child;
+    } else {
+      assert(_dragOffset != null);
```
Comment (dnfield) [inline packages/flutter/lib/src/material/refresh_indicator.dart]: This whole block should just be wrapped in an assert now\

Comment (dnfield) [reply packages/flutter/lib/src/material/refresh_indicator.dart]: (including the line below that github won't let me select)

**Review by dnfield — CHANGES_REQUESTED** [review]
Is there some benchmark that this will improve?  If not, we should write one before landing this.

If we don't have a benchmark for it, it will very easily get regressed (it will also be harder to tell if or how much this actually improves performance).

**Comment by LongCatIsLooong** [comment]
Hi @christian-muertz great job spotting this issue and thank you for the fix! However we can't merge the pull request without a valid test. I think a good test to add would be one that verifies `RefreshIndicator`'s child does not re-layout excessively. Are you still interested in working on this PR? I would be happy to work with you on adding a test!

**Comment by christian-muertz** [comment]
Yes iam still interested. I will take a look at it again on friday!

**Comment by LongCatIsLooong** [comment]
Hi @christian-muertz, any updates on this?

**Comment by christian-muertz** [comment]
No, i currently have no idea how to test whether the child was relayout.

**Comment by LongCatIsLooong** [comment]
Does something like this make sense (the code snippet makes use of some global variables in refresh_indicator_test.dart)?
```dart
  testWidgets(
    'RefreshIndicator does not force child to relayout',
    (WidgetTester tester) async {
      refreshCalled = false;
      int layoutCount = 0;

      Widget layoutCallback(BuildContext context, BoxConstraints constraints) {
        layoutCount += 1;
        return ListView(
          physics: const AlwaysScrollableScrollPhysics(),
          children: <String>['A', 'B', 'C', 'D', 'E', 'F'].map<Widget>((String item) {
              return SizedBox(
                height: 200.0,
                child: Text(item),
              );
          }).toList(),
        );
      };
      await tester.pumpWidget(
        MaterialApp(
          home: RefreshIndicator(
            onRefresh: refresh,
            child: LayoutBuilder(builder: layoutCallback),
          ),
        ),
      );

      await tester.fling(find.text('A'), const Offset(0.0, 300.0), 1000.0);
      await tester.pump();

      await tester.pump(const Duration(seconds: 1)); // finish the scroll animation
      await tester.pump(const Duration(seconds: 1)); // finish the indicator settle animation
      await tester.pump(const Duration(seconds: 1)); // finish the indicator hide animation
      expect(refreshCalled, true);
      expect(layoutCount, 1);
  });
```

**Comment by christian-muertz** [comment]
Yes it does make sense. I would suggest not checking for it to refresh because that is not what the test is for. I will add a test like this in a few mins :D

File: packages/flutter/lib/src/material/refresh_indicator.dart (hunk 1 above)
Comment (christian-muertz) [reply packages/flutter/lib/src/material/refresh_indicator.dart]: Makes sense

File: packages/flutter/lib/src/material/refresh_indicator.dart [hunk 2]
```diff
@@ -409,34 +409,33 @@ class RefreshIndicatorState extends State<RefreshIndicator> with TickerProviderS
...
-    assert(_dragOffset != null);
-    assert(_isIndicatorAtTop != null);
+    assert(() {
+      if (_mode == null) {
+        assert(_dragOffset == null);
+        assert(_isIndicatorAtTop == null);
+      } else {
+        assert(_dragOffset != null);
+        assert(_isIndicatorAtTop != null);
+      }
```
Comment (LongCatIsLooong) [inline packages/flutter/lib/src/material/refresh_indicator.dart]: I think there should be a `return true;` ?

Alternatively:
```dart
assert((_dragOffset == null) == (_mode == null));
assert((_isIndicatorAtTop == null) == (_mode == null));
```

Comment (christian-muertz) [reply packages/flutter/lib/src/material/refresh_indicator.dart]: You are right. I'll add a return true . I think its easier to understand !

**Review by LongCatIsLooong — APPROVED** [review]
LGTM.

File: packages/flutter/test/material/refresh_indicator_test.dart [hunk 3]
```diff
@@ -420,4 +420,39 @@ void main() {
     expect(controller.offset, lessThan(0.0));
     expect(refreshCalled, isTrue);
   }, variant: const TargetPlatformVariant(<TargetPlatform>{ TargetPlatform.iOS,  TargetPlatform.macOS }));
-}
+
+  testWidgets('RefreshIndicator does not force child to relayout', (WidgetTester tester) async {
+    int layoutCount = 0;
+
+    Widget layoutCallback(BuildContext context, BoxConstraints constraints) {
+      layoutCount++;
```
Comment (LongCatIsLooong) [inline packages/flutter/test/material/refresh_indicator_test.dart]: nit: our style guide prefers `+=` over `++`: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#prefer--over-

**Comment by LongCatIsLooong** [comment]
@dnfield it seems the change does prevent `RefreshIndicator` to re-layout its child when its state changes. Now that there's a test is a benchmark still required?

**Review by dnfield — APPROVED** [review]
LGTM

**Comment by dnfield** [comment]
Nope, this works too (we can test that we're not doing the extra work we don't need to do).

**Pull request merged**